hpx/runtime/serialization/access.hpp

```diff
@@ -76,15 +108,34 @@ namespace hpx { namespace serialization
 
             struct non_intrusive
             {
-                // this additional indirection level is needed to
-                // force ADL on the second phase of template lookup.
-                // call of serialize function directly from base_object
-                // finds only serialize-member function and doesn't
-                // perform ADL
+                template<class T1> struct dependent_false : std::false_type {};
+
+
                 template <class Archive>
                 static void call(Archive& ar, T& t, unsigned)
                 {
-                    detail::serialize_force_adl(ar, t, 0);
+                    if constexpr (has_serialize_adl<T>::value)
```

Comment:
Could you please guard the use of `if constexpr` with a feature test macro. The whole new serialization functionality should be made available when compiling in C++17 mode only anyways, so enabling `if constexpr` at the same time should be ok.

Reply:
`__cpp_if_constexpr` is not supported when compiling with a C++11 compiler. You might want to add a feature test here: https://github.com/STEllAR-GROUP/hpx/tree/master/cmake/tests, https://github.com/STEllAR-GROUP/hpx/blob/master/cmake/HPX_PerformCxxFeatureTests.cmake#L198, and here: https://github.com/STEllAR-GROUP/hpx/blob/master/cmake/HPX_AddConfigTest.cmake#L661. You can then use the preprocessor macro to hide the new code from older compilers.